shell/platform/darwin/macos/framework/Source/FlutterDisplayLink.mm

```diff
@@ -0,0 +1,356 @@
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterDisplayLink.h"
+
+#include "flutter/fml/logging.h"
+
+#include <algorithm>
+#include <optional>
+#include <thread>
+#include <vector>
+
+// Note on thread safety and locking:
+//
+// There are three mutexes used within the scope of this file:
+// - CVDisplayLink internal mutex. This is locked during every CVDisplayLink method
+//   and is also held while display link calls the output handler.
+// - DisplayLinkManager mutex.
+// - _FlutterDisplayLink mutex (through @synchronized blocks).
+//
+// Special care must be taken to avoid deadlocks. Because CVDisplayLink holds the
+// mutex for the entire duration of the output handler, it is necessary for
+// DisplayLinkManager to not call any CVDisplayLink methods while holding its
+// mutex. Instead it must retain the display link instance and then call the
+// appropriate method with the mutex unlocked.
+//
+// Similarly _FlutterDisplayLink must not call any DisplayLinkManager methods
+// within the @synchronized block.
+
+@class _FlutterDisplayLinkView;
+
+@interface _FlutterDisplayLink : FlutterDisplayLink {
+  _FlutterDisplayLinkView* _view;
+  std::optional<CGDirectDisplayID> _display_id;
+  BOOL _paused;
+}
+
+- (void)didFireWithTimestamp:(CFTimeInterval)timestamp
+             targetTimestamp:(CFTimeInterval)targetTimestamp;
+
+@end
+
+namespace {
+class DisplayLinkManager {
+ public:
+  static DisplayLinkManager& Instance() {
+    static DisplayLinkManager instance;
+    return instance;
+  }
+
+  void UnregisterDisplayLink(_FlutterDisplayLink* display_link);
+  void RegisterDisplayLink(_FlutterDisplayLink* display_link, CGDirectDisplayID display_id);
+  void PausedDidChange(_FlutterDisplayLink* display_link);
+  CFTimeInterval GetNominalOutputPeriod(CGDirectDisplayID display_id);
+
+ private:
+  void OnDisplayLink(CVDisplayLinkRef displayLink,
+                     const CVTimeStamp* inNow,
+                     const CVTimeStamp* inOutputTime,
+                     CVOptionFlags flagsIn,
+                     CVOptionFlags* flagsOut);
```

Comment:
nit: since we're in C++ code, use C++ identifiers like in_now, in_output_time, etc.